typing/uniqueness_analysis.ml

```diff
@@ -0,0 +1,1507 @@
+  (* it's still a tuple - we own it and nothing to do here *)
+  | Match_tuple _ -> UF.unused
+
+let mark_maybe_unique paths unique_use occ =
+  let mark_one path =
+    UF.singleton path (Maybe_unique (Maybe_unique.singleton unique_use occ))
+  in
+  UF.pars (List.map (fun path -> mark_one path) paths)
+
+(** returns:
+   the updated value.
+   the new introduced bindings.
+   usage during the process
+*)
+let pattern_match_var ~loc id value =
+  match value with
+  | Match_single { paths; _ } ->
+      (value, Ienv.Extension.singleton id paths, UF.unused)
+  | Match_tuple values ->
+      let path = UF.Path.fresh_root_of_ident id in
+      let ext = Ienv.Extension.singleton id [ path ] in
+      (* Mark all ps as seen, as we bind the tuple to a variable. *)
+      (* Can we make it aliases instead of used? Hard to do if we want
+         to preserve the tree-ness *)
+      ( Match_single { paths = [ path ]; loc },
+        ext,
+        UF.seqs
+          (List.map
+             (fun ({ paths; loc }, unique_use) ->
+               match unique_use with
+               | None -> UF.unused
+               | Some unique_use ->
+                   let occ = Occurrence.mk loc in
+                   mark_maybe_unique paths unique_use occ)
+             values) )
+
+(**
+handling pattern match of value against pat, returns ienv and uf.
+ienv is the new bindings introduced;
+uf is the usage caused by the pattern matching *)
+let rec pattern_match pat value =
+  match pat.pat_desc with
+  | Tpat_any -> (Ienv.Extension.empty, UF.unused)
+  | Tpat_var (id, _, _) ->
+      let _, ienv, uf = pattern_match_var ~loc:pat.pat_loc id value in
+      (ienv, uf)
+  | Tpat_alias (pat', id, _, _) ->
+      let value, ext0, uf0 = pattern_match_var ~loc:pat.pat_loc id value in
+      let ext1, uf1 = pattern_match pat' value in
+      (Ienv.Extension.conjunct ext0 ext1, UF.seq uf0 uf1)
+  | Tpat_constant _ ->
+      (Ienv.Extension.empty, mark_implicit_borrow_memory_address Read value)
+  | Tpat_tuple pats ->
+      pat_proj
+        ~handle_tuple:(fun values ->
+          (* We matched a tuple against a tuple parent and descend into each
+             case *)
+          (* no borrowing needed - we own the tuple! *)
+          let exts, ufs =
+            List.split
+              (List.map2
+                 (fun pat (value, _) -> pattern_match pat (Match_single value))
+                 pats values)
+          in
+          (Ienv.Extension.conjuncts exts, UF.seqs ufs))
+        ~extract_pat:Fun.id
+        ~mk_proj:(fun i _ -> Usage_tree.Projection.Tuple_field i)
+        value pats
+  | Tpat_construct (lbl, _, pats, _) ->
+      pat_proj ~extract_pat:Fun.id
+        ~mk_proj:(fun i _ ->
+          Usage_tree.Projection.Construct_field (Longident.last lbl.txt, i))
+        value pats
+  | Tpat_variant (lbl, mpat, _) ->
+      let uf = mark_implicit_borrow_memory_address Read value in
+      let t =
+        match value with
```

Comment:
This should just call the function that converts a `value_to_match` into a `single_value_to_match`.

Reply:
Resolved because we have taken a slight different approach.